src/ssram.c

```diff
@@ -3,40 +3,58 @@
 #include "hardware/dma.h"
 #include "hardware/gpio.h"
 #include "hardware/spi.h"
-
+#include "hardware/sync.h"
 #include "boards/pico_ice.h"
 
 #include "ice/ssram.h"
 
+#define CMD_WRITE              0x02
+#define CMD_READ               0x03
+
 static uint8_t g_dummy;
-static int g_tx_dma_channel = -1;
-static int g_rx_dma_channel = -1;
-static int g_irq;
+static volatile int g_tx_dma_channel = -1;
+static volatile int g_rx_dma_channel = -1;
+static volatile int g_irq;
+static volatile int g_cs_pin = -1;
+
+/// Wait that our own ongoing transaction completes.
+void ice_ssram_wait(void) {
+    while (ice_ssram_is_busy()) {
+        // Exiting the interrupt handler implicitly sets an event.
+        __wfe();
+    }
+}
+
+bool ice_ssram_is_busy(void) {
+    return g_cs_pin >= 0;
+}
 
 /// Select the SRAM for SPI transaction.
-/// Unlike regular SPI, PSRAM CS is active high.
-static void ice_ssram_select(void) {
-    // TODO: delay here to allow DRAM refresh?
+static void ice_ssram_select(int cs_pin) {
+    g_cs_pin = cs_pin;
+
+    // Short delay for two reasons: 1) to pull-up SS_SSRAM via pull-up resistor and
+    // 2) to ensure memory is deselected long enough to refresh DRAM.
+    sleep_us(1);
 
-    gpio_put(ICE_SSRAM_SPI_CS_PIN, true);
-    sleep_us(1); // We use a pullup on the board so it needs a bit of time to settle
+    gpio_put(cs_pin, false);
 }
 
 /// Select the SRAM for SPI transaction.
-/// Unlike regular SPI, PSRAM CS is active high.
 static void ice_ssram_deselect(void) {
+    assert(g_cs_pin >= 0);
```

Comment:
what about using an unsigned type here?

Reply:
It looks like the pico-c-sdk uses `uint` internally https://raspberrypi.github.io/pico-sdk-doxygen/group__hardware__gpio.html

Reply:
It's signed because it uses a negative value of g_cs_pin to signify that no memory chip is currently selected.